New audio library

rasky opened the pull request:
This PR introduces a new audio library, on top of audio.c.

This first step is a new RSP-based mixer. The mixer mixes and plays up to 32 independent audio streams called "waveforms". It also supports resampling: each waveform can play at a different playback frequency, which in turn can be different from the final output frequency. The resampling and mixing is performed by a very efficient RSP microcode (see rsp_mixer.S). Some figures on RSP mixer performances are on top of rsp_mixer.S, but for instance you can playback 32 channels at 44100Hz using ~20% of total RSP time (resampling is what takes time -- mixing itself would take less than 2% RSP).

The waveform API is the pluggable entry point to implement audio formats. For now, the only implemented waveform is "WAV64", a simple raw sample format. The implementation is trivial: samples are pulled from disk using `dma_read` when needed (well, I must say, it's trivial now that `dma_read` has been powered up to allow unaligned transfers...).

The whole mixer API and implementation has been carefully designed to be zero-copy: the CPU never ever touches the samples, no buffer is copied around. All samples are always directly generated, processed and mixed on the RSP, with the minimal amount of shuffling, and then directly send to AI via DMA (using the new `audio_write_begin` zero-copy API). There are only a few exceptions where we absolutely need to touch a few samples with CPUs every once in a while, but it's so rare and affects so few samples that performance is not impacted.

A new tool is introduced called "audioconv64". It converts trees of files into a destination folder, matching all known audio formats and converting them for the audio library. For now, the only supported conversion is `.wav` -> `.wav64`.

More PRs will follow to introduce other audio formats (no spoilers...).

rasky committed:
Add new audioconv64 tool.

This tool will be used for doing audio conversions for the audio library.
In this first iteration, we implement the main architecture with tree
traversal of input files, plus conversion of WAV files into the WAV64
format.

WAV files are parsed using the dr_wav single-file library, which allows
parsing and decoding must common WAV files included compressed ones,
for highest compatibility.

Ryzee119 commented on include/mixer.h:
```diff
@@ -0,0 +1,461 @@
+    float frequency;
+
+    // Length of the waveform, in number of samples. If the length is not
+    // known, this value can be initialized to a very large number (eg: 2**31-1)
```
As discussed on discord, 2^31-1 does not work (the sample will not play), currently the max len appears to be 0x40000 - 1.

rasky replied:
I've pushed a change that allows for longer waveforms to be played back, switching to 64-bit positions in C code while keeping RSP 32-bit (RSP just processes a short window at a time anyway).

Ryzee119 replied:
Thanks! Appears to work.
I changed my code to WAVEFORM_UNKNOWN_LEN with loop_len = 0 and had it running for the past 10 mins or so with no issues in the music stream.

Ryzee119 commented:
I can't comment on the subtle details on the N64 implementation, however I have incorporated this lib into my game port https://github.com/Ryzee119/cosmo64 with really good results.
Good work on this!
I tested the n64conv tool separately and the waveform conversion worked as expected.
I mainly tested generating PCM samples and feeding them in the mixer with a `waveformt_t` struct. Mixing music and sfxs upto 4 or 5 channels

meeq commented on examples/mixertest/Makefile:
```diff
@@ -0,0 +1,29 @@
+
+# Run audioconv on all WAV files under assets/
+# We do this file by file, but we could even do it just once for the whole
+# directory, becuase audioconv64 supports directory walking.
```
Typo: `because`

rasky replied:
fixed thanks

rasky committed:
mixer: add support for stereo waveforms.

Previously, the mixer only supported 1-channel waveforms. This commit
adds support for stereo waveforms, with interleaved samples. To support
this, the mixer needs to allocate two output channels (the specified
channel plus the following one), and the RSP ucode de-interleaves
the samples during the resampling step.

Calling all mixer APIs on the second channel of a stereo couple will
assert, as any configuration on that channel is actually invalid.

rasky commented:
I've added a commit that adds support for stereo waveforms (waveforms with interleaved left/right samples). Playing these waveforms use two mixer channels at the same time, as you need to mixing slots, but it's the best way in case your input source is already interleaved.

To avoid performance hits, this require a RSP ucode modification to de-interleave the left/right samples while resampling, in both 8-bit and 16-bit mode, but it's actually pretty easy otherwise and falls well within the overall design of the ucode.

snacchus commented on include/mixer.h:
```diff
@@ -0,0 +1,478 @@
+ * the RSP and store the final output stream directly in the final buffer which
+ * is sent via DMA to the N64 AI hardware DAC. In general, the N64 CPU is very
+ * bandwidth-limited on RDRAM accesses, so "touching" 44100 samples per second
+ * with the PCU would already consume too many resources. Audio players are
```
PCU -> CPU

snacchus commented on include/mixer.h:
```diff
@@ -0,0 +1,478 @@
+ * If the waveform is marked as stereo (channels == 2), the mixer will need
+ * two channels to play it back. "ch" will be used for the left samples,
+ * while "ch+1" will be used for the right samples. After this, it is
+ * forbidden
```
Is there some text missing here?

snacchus commented on include/mixer.h:
```diff
@@ -0,0 +1,478 @@
+ * whenever a new unavailable portion of the waveform is requested.
+ *
+ * "wpos" indicates the absolute position in the waveform from which
+ * starts reading, and "wlen" indicates how many samples to read.
```
"...from which _it_ starts reading,..."?

snacchus commented on include/mixer.h:
```diff
@@ -0,0 +1,478 @@
+ *     already partially cached in the sample buffer's channel).
+ *   * At user request (#mixer_ch_set_pos).
+ *   * At loop: if a loop is specified in the waveform (loop_len != 0), the
+ *     mixer will seek when requires to execute the loop.
```
requires -> required

snacchus commented on include/mixer.h:
```diff
@@ -0,0 +1,478 @@
+    // Name of the waveform (for debugging purposes)
+    const char *name;
+
+    // Width of a sample of this waveform, in bits. Supported values are 8 or 16.
+    // Notice that samples must always be signed.
+    uint8_t bits;
+
+    // Number of interleaved audio channels in this waveforms. Supported values
+    // are 1 and 2 (mono and stereo waveforms). Notice that a stereo waveform
+    // will use two consecutive mixer channels to be played back.
+    uint8_t channels;
```
Maybe use enums or defines for supported values? Like `WAVEFORM_SAMPLE_WIDTH_16` or `WAVEFORM_CHANNELS_MONO`

rasky replied:
I'd rather keep them numeric because I use their actual numeric value in formulas (eg: `num_samples * channels`).

snacchus commented on include/wav64.h:
```diff
@@ -0,0 +1,59 @@
+ * but rather loaded on request when needed for playback, streaming directly
+ * from ROM. See #waveform_t for more details.
+ * 
+ * Use #wav64_play to playback. For more advances usage, call directly the
```
advances -> advanced

snacchus commented on include/mixer.h:
```diff
@@ -0,0 +1,478 @@
+
+    // Width of a sample of this waveform, in bits. Supported values are 8 or 16.
+    // Notice that samples must always be signed.
+    uint8_t bits;
```
In some places the sample width is defined in bytes and sometimes in bits. Is there a specific reason for this? If not, I would make it consistent, probably in favor of bytes.

rasky replied:
The general idea is to use bits for the public API, because I think everybody is familiar with the concept of audio being either 8-bit or 16-bit (I think the sentence "let's play this 8-bit wav" is more clear than "let's play this 1-byte wav"). 

But then internally, bits are not very useful and I mainly need two different things: "bytes per sample" (that also includes the number of channels into account, so it is something like `bits / 8 * channels`) and "bitshifts per samples", which is the shift amount that you can use to convert sample indices to byte offsets viceversa. For instance, a 16-bit stereo audio requires 4 bytes per sample, so the shift amount to use is `<< 2`. Obviously you can't really map everything conceivable into a shift, but as long as we are limited to 8 or 16 bits and 1 or 2 channels, then you can indeed do everything with shifts, which makes the code faster.

The issue with this is that in mixer.c and samplebuffer.c unfortunately I'm using the "bps" acronym for both "bytes per sample" and "bitshift per sample", which can make the code a little confusing to read. I could fix this in the future to help readability.

But going back to your comment, I think that it's the correct choice to use bits in public API and bytes internally. It might be less uniform, but I think it's clearer for users.

snacchus replied:
> I think that it's the correct choice to use bits in public API and bytes internally. It might be less uniform, but I think it's clearer for users.

That's fair enough. I was thinking about the API more as a programmer I guess 😄

Ryzee119 commented on src/audio/wav64.c:
```diff
@@ -0,0 +1,79 @@
+}
+
+void wav64_set_loop(wav64_t *wav, bool loop) {
+	wav->wave.loop_len = wav->wave.len;
```
never checks `bool loop`
Should this be something like `wav->wave.loop_len = (loop) ? wav->wave.len : 0;`

So you can do `wav64_set_loop(sfx, false)` to disable looping.

rasky replied:
Good catch, I'll fix that.

snacchus commented on tools/audioconv64/audioconv64.c:
```diff
@@ -0,0 +1,171 @@
+#if __BYTE_ORDER__ == __ORDER_BIG_ENDIAN__
+#define LE32_TO_HOST(i) (((uint32_t)((i) & 0xFF000000) >> 24) | ((uint32_t)((i) & 0x00FF0000) >>  8) | ((uint32_t)((i) & 0x0000FF00) <<  8) | ((uint32_t)((i) & 0x000000FF) << 24))
+#define BE32_TO_HOST(i) (i)
+#define HOST_TO_LE32(i) BE32_TO_HOST(i)
```
Should probably be `LE32_TO_HOST(i)` right?

snacchus commented on tools/audioconv64/audioconv64.c:
```diff
@@ -0,0 +1,171 @@
+#define LE16_TO_HOST(i) (i)
+#define HOST_TO_BE16(i) ((((i) & 0xFF00) >> 8)  | (((i) & 0xFF) << 8))
+#define HOST_TO_LE16(i) (i)
+#endif
```
Some expressions are duplicated here. Maybe refactor them out?

rasky replied:
Done

snacchus commented on src/audio/rsp_mixer.S:
Is it still outdated?

rasky replied:
Yes. I've fixed It now.

snacchus commented on src/audio/rsp_mixer.S:
These zero-filled data sections could be moved to .bss as well.

rasky replied:
Done, but not for WAVFORM_SETTINGS which is input data, and I don't want to move it to a different section where its address might change.

snacchus commented on src/audio/rsp_mixer.S:
Maybe do this at the start of the loop?

rasky replied:
Done

snacchus commented on src/audio/rsp_mixer.S:
There is a comment at the start of the function about what `k1` is, but maybe add a define for it anyway, because it will make the code much more readable. This generally applies to function arguments.

rasky replied:
Done. But I can't do that for function arguments in general because they're function-relative, so I can't easily add defines to the calling sites.

I think maybe the DMA function are the one creating more issues here. I think those might have to be learnt by heart but in general s0/s4 are often input/output, and t0 is count. t2=pitch (not used in this microcode) is possibly the odd one. If you see Mixer, it also takes s4 as its output argument.

snacchus replied:
That's fair. Also see my previous comment.

snacchus commented on src/audio/rsp_mixer.S:
The comment says `<= 8` but this actually checks for `< 8`

rasky replied:
Good catch, I'll fix this

snacchus reviewed (commented):
I think this is some fantastic work! The API is very clean, well designed and thouroughly documented. I just left some comments on spelling mistakes or very small issues here and there. Since I don't know very much about audio processing, I can't really comment on the correctness of the implementation. But I trust that it is sound since it has already been tested.

The ucode seems good as well, apart from the few comments I left there is one thing maybe: I have found that ucode (and assembly in general of course) becomes much more readable and easy to follow if you give meaningful names to registers via defines and such. Of course if they are just scratch registers used over a couple lines that's fine, but as their scope becomes longer, it gets more complicated to follow the program's logic. The same applies to function arguments. I'm not sure how to handle passing arguments in from outside though. As in, I'm calling a function with `jal` and need to set the arguments, you still need to scroll all the way to the function definition to know the argument's meanings. Maybe we can come up with a scheme for this. Maybe this is just me who hasn't developed in asm for very long, but I think it will be better for newcomers as well.

The only other concern I have is the lack of an asynchronous API. Something like `mixer_poll_start()` and `mixer_poll_wait()`. Because having the CPU stall while the RSP is running seems like a waste. This becomes even more important once we merge the new kernel. Or was your intention to move `mixer_poll` into the audio thread anyway and handle asynchrony on the RSP-API level?

Once again, this is some great work, well done!

rasky commented:
> The ucode seems good as well, apart from the few comments I left there is one thing maybe: I have found that ucode (and assembly in general of course) becomes much more readable and easy to follow if you give meaningful names to registers via defines and such. Of course if they are just scratch registers used over a couple lines that's fine, but as their scope becomes longer, it gets more complicated to follow the program's logic.

Yes I agree. I think the resampling and mixing loops are pretty in line with what you describe, but the main loop at the start of the ucode doesn't use many defines. I'll change that.

> The same applies to function arguments. I'm not sure how to handle passing arguments in from outside though. As in, I'm calling a function with jal and need to set the arguments, you still need to scroll all the way to the function definition to know the argument's meanings. Maybe we can come up with a scheme for this.

I must say that it's not much different in C; if I see a function call point, I don't know the argument meanings without going to the function definition. Maybe the difference is that it's less clear which registers are used as function arguments?

> The only other concern I have is the lack of an asynchronous API. Something like mixer_poll_start() and mixer_poll_wait(). Because having the CPU stall while the RSP is running seems like a waste. This becomes even more important once we merge the new kernel. Or was your intention to move mixer_poll into the audio thread anyway and handle asynchrony on the RSP-API level?

Yeah the intention was to run mixer_poll into the audio thread, and let it yield while the RSP is running. An async API can be designed of course but it wouldn't be so easy to use. Consider that mixer_poll can call into RSP multiple times, so you would have to change the API so that each call to mixer_poll_wait returns the number of samples actually generated, so that the caller can keep on repeating calls until the required number of samples is reached. On top of that, consider that waveform_t reading functions might use RSP as well, for instance to decompress VADPCM or MPEG2. It's even less clear how we would handle that in an async API.

rasky committed:
mixer: improve comments and readibility of ucode register names

Also move a couple of arrays from data to bss

rasky committed:
mixer: fix a couple of pessimizations bugs in the 8-channel mixer

1) The 8-channel mixer core was not called for exactly 8
channels but only up to 7.

2) The last iteration of the 8-channel mixer core would run as a
32-channel mixer.

rasky committed:
wav64: fix implementation of wav64_set_loop to allow disabling loops

snacchus commented:
> I must say that it's not much different in C; if I see a function call point, I don't know the argument meanings without going to the function definition. Maybe the difference is that it's less clear which registers are used as function arguments?

Very true. In my ucode I experimented with adding global defines for function arguments with the function name as prefix. So for example `DmaDramAddr` or something. But obviously that only works if the function is defined before it is ever used, which is not always possible and it also clutters the global "namespace". I can't think of anything else so I think this is fine for now.

> Yeah the intention was to run mixer_poll into the audio thread, and let it yield while the RSP is running. An async API can be designed of course but it wouldn't be so easy to use. Consider that mixer_poll can call into RSP multiple times, so you would have to change the API so that each call to mixer_poll_wait returns the number of samples actually generated, so that the caller can keep on repeating calls until the required number of samples is reached. On top of that, consider that waveform_t reading functions might use RSP as well, for instance to decompress VADPCM or MPEG2. It's even less clear how we would handle that in an async API.

If `mixer_poll` is always put into the audio thread then that's fine by me. It should work pretty well that way. The issues you mentioned with an async API are valid, and I hadn't thought of them. It would make the API way too confusing so it's probably not a good idea.

meeq commented on src/audio/mixer.c:
```diff
@@ -0,0 +1,645 @@
+	Mixer.vol = vol;
+}
+
+void mixer_close(void) {
```
`mixer_close` does not appear in `mixer.h` -- was this intentional?

rasky replied:
No it's an oversight. I'll add it, thanks.

meeq commented on src/audio/mixer.c:
```diff
@@ -0,0 +1,645 @@
+	Mixer.vol = vol;
+}
+
+void mixer_close(void) {
```
`mixer_close` does not appear in `mixer.h` -- was this intentional?

meeq commented:
I'm trying to build the tools on this branch, and I'm getting this cryptic error:

```
make -C audioconv64
make[1]: Entering directory 'libdragon/tools/audioconv64'
cc -std=gnu11 -MMD -O2 -Wall -Wno-unused-result -Werror -I../../include  -o audioconv64 audioconv64.c
In file included from audioconv64.c:38:
conv_wav64.c: In function ‘wav_convert’:
conv_wav64.c:44:9: error: ‘strncpy’ output truncated before terminating nul copying 4 bytes from a string of the same length [-Werror=stringop-truncation]
   44 |         strncpy(head.id, "WV64", 4);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[1]: *** [Makefile:7: audioconv64] Error 1
```

I'm using GCC 11.1.0 on x64 Manjaro Linux

**EDIT**

I was able to work around this with the following patch:
```diff
diff --git a/tools/audioconv64/conv_wav64.c b/tools/audioconv64/conv_wav64.c
index 4fbaf2d..f4699be 100644
--- a/tools/audioconv64/conv_wav64.c
+++ b/tools/audioconv64/conv_wav64.c
@@ -38,18 +38,17 @@ int wav_convert(const char *infn, const char *outfn) {
 		loop_len -= 1;
 	}
 
-	wav64_header_t head;
-	memset(&head, 0, sizeof(wav64_header_t));
-
-	strncpy(head.id, "WV64", 4);
-	head.version = WAV64_FILE_VERSION;
-	head.format = 0;
-	head.channels = wav.channels;
-	head.nbits = nbits;
-	head.freq = HOST_TO_BE32(wav.sampleRate);
-	head.len = HOST_TO_BE32(cnt);
-	head.loop_len = HOST_TO_BE32(loop_len);
-	head.start_offset = HOST_TO_BE32(sizeof(wav64_header_t));
+	wav64_header_t head = {
+		.id = {'W', 'V', '6', '4'},
+		.version = WAV64_FILE_VERSION,
+		.format = 0,
+		.channels = wav.channels,
+		.nbits = nbits,
+		.freq = HOST_TO_BE32(wav.sampleRate),
+		.len = HOST_TO_BE32(cnt),
+		.loop_len = HOST_TO_BE32(loop_len),
+		.start_offset = HOST_TO_BE32(sizeof(wav64_header_t)),
+	};
 
 	if (flag_verbose)
 		fprintf(stderr, "Converting: %s => %s\n", infn, outfn);
```

Ryzee119 commented:
> conv_wav64.c:44:9: error: ‘strncpy’ output truncated before terminating nul copying 4 bytes from a string of the same length [-Werror=stringop-truncation]
>    44 |         strncpy(head.id, "WV64", 4);
>       |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~

What if a memcpy is used instead of a strncpy?

This is a very pendandic error from -Wall and -Werror. Otherwise we need something like `-Wno-error=stringop-truncation`?

meeq commented:
I have [replaced the mixer in FlappyBird with the implementation from this branch](https://github.com/meeq/FlappyBird-N64/commit/4bd5978bcb258ca70d944aada52583c21d0e75a0), but audio isn't working for me in cen64 or m64p.

Compiled ROM: [FlappyBird.zip](https://github.com/DragonMinded/libdragon/files/7298156/FlappyBird.zip)


There isn't much surface area to screw up here, but I must be doing something wrong.

Ryzee119 commented:
> I have [replaced the mixer in FlappyBird with the implementation from this branch](https://github.com/meeq/FlappyBird-N64/commit/4bd5978bcb258ca70d944aada52583c21d0e75a0), but audio isn't working for me in cen64 or m64p.
> 
> Compiled ROM: [FlappyBird.zip](https://github.com/DragonMinded/libdragon/files/7298156/FlappyBird.zip)
> 
> There isn't much surface area to screw up here, but I must be doing something wrong.

Tested your zip and have audio here (With Cen64) 🤔 Cool game btw!

rasky committed:
rsp_mixer: fix condition of DMA end

This is just a very slight optimization: we can use the unrolled loop
also in the case in which pos+step*8 reaches exactly DMA end. Before,
there was an off-by-one error which fell back to the normal loop in
this edge case.

rasky commented:
@meeq you probably need to update emulators to the latest version. The new dma_read() function which works with misaligned addresses requires at least a partial implementation of PI DMA undocumented behaviours, which has been done on cen64, m64p, and ares.

Notice that I'm not a fan of breaking emulators, I thought *a lot* of this. But for audio and in general, the advantage of unlocking this previously unknown potential of the PI DMA is very useful, and hopefully the emulators will catch up in a short time. Also I'm personally opening PRs for this to make up with having broken them :)

rasky committed:
audioconv64: avoid using strncpy which triggers security warnings on compilers

rasky commented:
> I'm trying to build the tools on this branch, and I'm getting this cryptic error:

thanks, just pushed the fix for this to this PR

rasky committed:
mixer: add missing mixer_close prototype

rasky committed:
audioconv64: simplify byte order macros

rasky commented:
Ok I think I've applied all review comments. I think nobody has actually reviewed the C implementation yet, but I understand that not many people are familiar with low level digital audio processing.

On the other hand, it's been tested in three different games with good results, and I have done endless tests on it so I think the implementation is indeed correct (I've just fixed some small glitch with stereo waveforms -- but those were only recently introduced). I believe the implementation is mostly sound.

@anacierdem for me it would be good to merge. If you think you will have time to do an implementation review in the next few days we can hold it, otherwise I think I'd rather get it in now so that we can get on with XM player (which is hopefully a much simpler review).

rasky committed:
rsp_mixer: fix typo in define name and add a couple of register aliases

rasky commented:
I realized I hadn't updated the example Makefile to latest bells & whistles, nor the CI failed because I forgot to add the Makefile to the examples/Makefile. I have fixed it now and verified the CI compiles audioconv64 and then the mixertest correctly, with asset conversion.

meeq commented on tools/Makefile:
```diff
@@ -1,13 +1,14 @@
+	$(MAKE) -C audioconv64 install
 
 .PHONY: clean
 clean:
```
`make clean` does not clean the `audioconv64` sub-folder

rasky replied:
thanks, fixed

Ryzee119 commented on examples/mixertest/mixertest.c:
```diff
@@ -0,0 +1,93 @@
+        // frame to perform mixing.
+        if (audio_can_write()) {    	
+	        short *buf = audio_write_begin();
+			mixer_poll(buf, audio_get_buffer_length());
```
Tabs/Spaces

rasky replied:
Done

Ryzee119 commented on src/audio/mixer.c:
```diff
@@ -0,0 +1,646 @@
+	uint32_t flags;
+} mixer_channel_t;
+
+typedef struct rsp_mixer_channel_s {
```
There's an assert on the `rsp_mixer_channel_s` struct size. Do we want to use packed to make this explicit?
As they're all 4 byte data datatypes it shouldnt really be an issue, I wouldnt hold anything up with this.

rasky replied:
Done. I don't think it matters per-se, as long as the RSP definition is the same. That is, even if the non-packed structure had a uint8_t, it would be sufficiente for the RSP DMEM structure to follow it with its packing. OTOH, RSP allows misaligned memory accesses, so packed saves DMEM, doesn't slow down the RSP, and doesn't force the programmer to think of padding rules. So all in all I think you're right that it's better.

Ryzee119 commented on src/audio/mixer.c:
```diff
@@ -0,0 +1,646 @@
+
+	// Do one large allocations for all sample buffers
+	assert(Mixer.ch_buf_mem == NULL);
+	Mixer.ch_buf_mem = malloc(totsize);
```
assert after this malloc to confirm it was successful?

rasky replied:
Done

Ryzee119 commented on src/audio/mixer.c:
```diff
@@ -0,0 +1,646 @@
+
+	tracef("mixer_exec: 0x%x samples\n", num_samples);
+
+    uint32_t fake_loop = 0;
```
tabs/spaces

rasky replied:
Done

Ryzee119 reviewed (commented):
Sorry for the late review but I flicked over the code, which i hadnt really done previously when I was just testing the functionality.

Looks really good to me. I only found a couple things, all minor and mostly nits.

rasky committed:
Add mixer example.

Add a simple example that shows basic usage of mixer with WAV64 files.
Assets are converted at build time via audioconv64 to show how
to use the tool.

rasky committed:
mixer: allow to playback waveforms with more then 0x40000 samples.

Currently, mixer is limited to short waveforms because the RSP uses
internally a 1.19.12 fixed point format to do resampling, so you can't
represent a byte position bigger than 2^19 (which is 2^18 for 16-bit
samples, and even 2^17 for future support of stereo waveforms). That's
too little.

This PR changes the mixer to keep internally 64-bit fixed pointer
numbers (1.51.12), and then truncate them properly when passing them
to RSP. RSP in fact doesn't need to compute across the whole waveform
but can work only with a smaller window.

rasky committed:
audio: add new RSP-based mixer

This module offers a flexible API to mix and play up to 32 independent
audio streams called "waveforms". It also supports resampling: each
waveform can play at a different playback frequency, which in turn can
be different from the final output frequency. The resampling and mixing
is performed by a very efficient RSP microcode (see rsp_mixer.S).

The mixer exposes 32 channels that can be used to play different audio
sources. An audio source is called a "waveform", and is represented by
the type waveform_t. To be able to produce audio that can be mixed
(eg: decompress and playback a MP3 file), the decoder/player code must
implement a waveform_t.

Currently, the only implemented waveform is WAV64, which is a simple raw
audio format.

The following commits will add a conversion tool for WAV files, and an
example on how to use the mixer.
